src/Microsoft.Health.Fhir.Api/Controllers/AadProxyController.cs

```diff
@@ -0,0 +1,363 @@
+        /// <param name="sessionState">session_state URL parameter.</param>
+        /// <param name="error">error URL parameter.</param>
+        /// <param name="errorDescription">error_description URL parameter.</param>
+        [HttpGet("callback/{encodedRedirect}/{launchContext}")]
+        public ActionResult Callback(
+            string encodedRedirect,
+            string launchContext,
+            [FromQuery(Name = "code")] string code,
+            [FromQuery(Name = "state")] string state,
+            [FromQuery(Name = "session_state")] string sessionState,
+            [FromQuery(Name = "error")] string error,
+            [FromQuery(Name = "error_description")] string errorDescription)
+        {
+            Uri redirectUrl = new Uri(Base64Decode(encodedRedirect));
+
+            if (!string.IsNullOrEmpty(error))
+            {
+                return RedirectPermanent($"{redirectUrl.ToString()}?error={error}&error_description={errorDescription}");
+            }
+
+            string compoundCode;
+            try
+            {
+                JObject launchParameters = JObject.Parse(Base64Decode(launchContext));
+                launchParameters.Add("code", code);
+                compoundCode = Uri.EscapeDataString(Base64Encode(launchParameters.ToString(Newtonsoft.Json.Formatting.None)));
+            }
+            catch
+            {
+                _logger.LogError("Error parsing launch parameters.");
+                throw;
+            }
+
+            return RedirectPermanent($"{redirectUrl.ToString()}?code={compoundCode}&state={state}&session_state={sessionState}");
+        }
+
+        /// <summary>
+        /// Proxies a (POST) request to the AAD token endpoint
+        /// </summary>
+        /// <param name="grantType">grant_type request parameter.</param>
+        /// <param name="compoundCode">The base64 encoded code and launch context</param>
+        /// <param name="redirectUri">redirect_uri request parameter.</param>
+        /// <param name="clientId">client_id request parameter.</param>
+        /// <param name="clientSecret">client_secret request parameter.</param>
+        [HttpPost("token")]
+        public async Task<ActionResult> Token(
+            [FromForm(Name = "grant_type")] string grantType,
+            [FromForm(Name = "code")] string compoundCode,
+            [FromForm(Name = "redirect_uri")] Uri redirectUri,
+            [FromForm(Name = "client_id")] string clientId,
+            [FromForm(Name = "client_secret")] string clientSecret)
+        {
+            EnsureArg.IsNotNull(grantType, nameof(grantType));
+            EnsureArg.IsNotNull(clientId, nameof(clientId));
+
+            var client = _httpClientFactory.CreateClient();
+
+            // TODO: Add handling of 'aud' -> 'resource'
+            if (grantType != "authorization_code")
+            {
+                List<KeyValuePair<string, string>> fields = new List<KeyValuePair<string, string>>();
```

Comment:
Quick question: am I right in understanding this is a guaranteed failure condition?

If so, why not just return an "unsupported grant type" error? (And if not, which other grant types do we expect here?)

Reply:
Actually, that is not a guaranteed failure. AAD supports client_credentials, etc. and we don't want to "disable" those features just because we enable SMART proxy. It would actually cause us a lot of problems if we did because many of the E2E tests obtain tokens that way. So this is sort of a bypass; if it is not authorization code, just pass on to AAD>

Reply:
I think we should change the tests to use an environment variable instead and make this case an error as Josh suggests. Not in this PR though...

Reply:
@johnstairs, I agree, I will put a TODO note in the code to remove this. I think we have will some code changes related to multiple IDPs that will affect this too. So I will mark the code for now.